packages/sanity-plugin-nacelle-input/src/components/NacelleLinker.js

```diff
@@ -0,0 +1,307 @@
+          </Select>
+        )}
+        {multiTab && (
+          <TabList className="tab">
+            {dataTypes.map((type, idx) => (
+              <Tab
+                key={type}
+                label={type}
+                selected={idx === activeTab}
+                className="tablinks"
+                onClick={() => setActiveTab(idx)}
+                space={2}
+              />
+            ))}
+          </TabList>
+        )}
+      </Stack>
+      <TabPanel>{children}</TabPanel>
+    </Box>
+  )
+}
+
+Interface.propTypes = {
+  dataType: PropTypes.oneOfType([PropTypes.array, PropTypes.string]),
+  interfaceOpen: PropTypes.bool.isRequired,
+  activeTab: PropTypes.number,
+  setActiveTab: PropTypes.func,
+  children: PropTypes.node
+}
+
+const NacelleLinker = ({ type, onChange, value, markers, level, readOnly }) => {
+  const [searchQuery, setSearchQuery] = useState(null)
+  const [spaceOptions, setSpaceOptions] = useState(null)
+  const [activeTab, setActiveTab] = useState(0)
+  const [interfaceOpen, setInterfaceOpen] = useState(false)
+  const onClose = useCallback(() => setInterfaceOpen(false), [])
+  const onQueryUpdate = useCallback((query) => {
+    setSearchQuery(query)
+  }, [])
+
+  useEffect(() => {
+    if (!spaceOptions) {
+      const initialSpace = Array.isArray(config.nacelleSpaces)
+        ? config.nacelleSpaces.find(
+            (s) => s.nacelleEndpoint && s.nacelleSpaceToken && s.spaceName
+          )
+        : {}
+      setSpaceOptions(initialSpace)
+    }
+  }, [spaceOptions])
+
+  const handle = value || ''
+
+  const inputId = useId()
+
+  const selectItem = (handle) => {
+    onChange(createPatchFrom(handle))
+    onClose()
+  }
+
+  const dataTypeFromOptions = type.options && type.options.dataType
+  let dataType
+
+  if (!dataTypeFromOptions) {
+    dataType = ['collections', 'products']
+  } else {
+    dataType = Array.isArray(dataTypeFromOptions)
+      ? dataTypeFromOptions
+      : [dataTypeFromOptions]
+  }
+
+  return (
+    <ThemeProvider theme={studioTheme}>
+      {interfaceOpen && (
+        <Dialog
+          header="Indexed PIM Data"
+          id="dialog-example"
+          onClose={onClose}
+          width={1}
+          zOffset={1000}
+        >
+          <HandleContext.Provider value={{ handle, setHandle: selectItem }}>
+            <SearchQueryContext.Provider
+              value={{ searchQuery, setSearchQuery }}
+            >
+              <SpaceOptionsContext.Provider
+                value={{ spaceOptions, setSpaceOptions }}
+              >
+                <Interface
+                  dataType={dataType}
+                  interfaceOpen={interfaceOpen}
+                  activeTab={activeTab}
+                  setActiveTab={setActiveTab}
+                >
+                  <TextInput
+                    fontSize={[2, 2, 3]}
+                    icon={SearchIcon}
+                    placeholder="Search indexed entries"
+                    onChange={(event) =>
```

Comment:
We might want to consider debouncing the query updates here. Currently this fires on every keystroke, which triggers data fetching per keystroke. Would likely give users a smoother experience. Could be future work?

Reply:
I think debouncing this makes a lot of sense